### ISSUE TITLE
make textandversion (recoverabletextandversion) and treeandversion (recoverable tree) works same way

### ISSUE THREAD

**Issue by heejaechang** [dotnet/roslyn#1157]
currently two are similar but the way it works are completely different. and TextAndVersion has an issue where just asking for a version requires having full text. this is especially a big perf issue when some body is asking text version for a forked document.

**Comment by heejaechang** [comment]
@jaredpar @pharring opened a bug for the perf issue.

**Comment by heejaechang** [comment]
related code
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/RecoverableTextAndVersion.cs,f263730f33004943
http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs,eb5b0e13397f84a5
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/TextAndVersion.cs,58c9881cf69c5231
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/TreeAndVersion.cs,5b2ba719fc4e5f53
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/DocumentState.cs,cb2652a0e852701f

basically, for tree, tree itself is either recoverable or not and TreeAndVersion is independent from recover-ability.

but for text, whole TextAndVersion is whether recoverable or not, and two can't be separated. and RecoverableTextAndVersion and regular TextAndVersion doesn't even share same base type.

**Comment by Pilchie** [comment]
@mattwar - do you want to take a look at this?  I know you had some concerns about the complexity of the way `SyntaxTree` does recovery.

**Comment by jaredpar** [comment]
Dug into this problem a bit deeper today.  The root issue is in [CreateRecoverableTextAndTree](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/DocumentState.cs,cb2652a0e852701f):

``` csharp
            // this captures the lazyTree local
            var lazyText = new AsyncLazy<TextAndVersion>(
                c => GetTextAndVersionAsync(lazyTree, textVersion, encoding, filePath, c),
                c => GetTextAndVersion(lazyTree, textVersion, encoding, filePath, c),
                cacheResult: false);
```

Notice in particular: `cacheResult: false`.  This means any use of the `AsyncLazy<TextAndVersion>`, sync or async, will completely reconstruct the text of the tree using `SyntaxNode.WriteTo`.  There is no caching involved here, every single use will reconstruct the tree.  

The `AsyncLazy<TextAndVersion>` eventually becomes `TextDocumentState.textSource`.  This is where calls like `Document.GetSyntaxVersion` end up bottoming out at and hence also end up always rebuilding the text of the tree.

**Comment by heejaechang** [comment]
cacheResult:false is one thing. it will reduce it to happen once, but in this particular case (forked document), it probably doesn't need to happen at all if TextAndVersion works like TreeAndVersion where Text itself is recoverable like tree. then, asking version won't require text being realized.

**Comment by jaredpar** [comment]
This has a trickle down effect on the IDE code.  It means that creating items like `SyntaxTreeIdentifierInfo` will both recreate the `SyntaxTree` and the text involved.

**Comment by jaredpar** [comment]
@heejaechang why not make getting the version independent of the text?  Basically make it a three way tuple (text, tree and version) instead of a two way tuple (text + version and tree).  

Even if the text is recoverable it is still being unnecessarily created.  The motivating case here is the code formatter tool.  There are only two times where the text of the tree needs to be realized:

- When parsing out the initial trees.
- When writing the final `SyntaxTree` values to disk. 

The tool itself never looks at the text.  Hence if the text is created it's done by the IDE and seems very wasteful.  This is doubly true when we run against large solutions like Roslyn.

**Comment by heejaechang** [comment]
I think that's basically what I am saying. Textandversion and treeandversion becomes asynclazy and text and tree itself in text/treeversion also become lazy. (here two versions are different so can't be shared)

What I described above is already how treeandversion works.

Sent from my Windows Phone
________________________________

**Comment by mattwar** [comment]
TextAndVersion is a public API. Changing it to have the text be lazy would be a breaking change, and have lots of downstream effects.  You could, however, introduce a different type internally that has different characteristics.

**Comment by mattwar** [comment]
It is perfectly reasonable to change how DocumentState stores this so that version is accessable without causing the text to be recomputed.
